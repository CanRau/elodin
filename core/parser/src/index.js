import Parser from './Parser'
import parse from './parse'
import errorTypes from './errorTypes'

export { Parser, parse, errorTypes }
