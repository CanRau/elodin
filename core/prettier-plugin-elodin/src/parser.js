import { format } from '@elodin/format'

export default function parse(text, _, options) {
  const formatted = format(text)

  const errors = formatted.errors
  let formattedText = formatted.code

  if (errors.length > 0) {
    let formattedText = text
  }

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    error: errors[0],
    end: text.length,
    source: text,
    start: 0,
  }
}
