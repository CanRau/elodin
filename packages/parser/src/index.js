import Parser from './Parser'
import parse from './parse'

export { Parser, parse }
