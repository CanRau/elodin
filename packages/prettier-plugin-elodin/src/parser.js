import { format } from '@elodin/format'

export default function parse(text) {
  const formattedText = format(text)

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    end: text.length,
    source: text,
    start: 0,
  }
}
