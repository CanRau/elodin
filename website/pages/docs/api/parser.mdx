import { make as DocLayout } from '../../../components/DocLayout.bs.js'

export default DocLayout

# Parser

`@elodin/parser`

The parser turns a string of Elodin code into an AST (Abstract Syntax Tree).

## parse

It exposes a `parse` method that just takes an single string of code.

### Example

```javascript
import { parse } from '@elodin/parser'

const input = `
view Button {
  backgroundColor: red
  paddingLeft: 10
  paddingRight: 10
}
`

const ast = parse(input)
```

## AST Nodes

The AST consists of the following AST nodes:

### File

| Property | Type | Description |
| -------- | ---- | ----------- |
| body     |      |             |

### Style

| Property | Type | Description |
| -------- | ---- | ----------- |
| name     |      |             |
| format   |      |             |
| body     |      |             |
| comments |      |             |

### Variant

| Property | Type | Description |
| -------- | ---- | ----------- |
| name     |      |             |
| body     |      |             |
| comments |      |             |

### Conditional

| Property | Type | Description |
| -------- | ---- | ----------- |
| property |      |             |
| value    |      |             |
| operator |      |             |
| boolean  |      |             |
| body     |      |             |

### Declaration

| Property | Type | Description |
| -------- | ---- | ----------- |
| property |      |             |
| value    |      |             |
| dynamic  |      |             |
| raw      |      |             |

### FunctionExpression

| Property | Type | Description |
| -------- | ---- | ----------- |
| callee   |      |             |
| params   |      |             |

### Float

| Property   | Type | Description |
| ---------- | ---- | ----------- |
| integer    |      |             |
| fractional |      |             |
| negative   |      |             |

### Integer

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
| negative |      |             |

### Identifier

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

### String

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

### Variable

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
