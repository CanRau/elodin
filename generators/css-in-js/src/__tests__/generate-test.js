import { parse } from '@elodin/parser'

import { createGenerator } from '../index'

import felaAdapter from '../adapters/fela'
import reactFelaAdapter from '../adapters/react-fela'

const file = `
variant Type {
  Primary
  Secondary
}

variant Mode {
  Dark
  Light
}

view Button {
  __border: 0
  backgroundColor: red
  paddingLeft: 10
  borderWidth: $borderWidth
  [Type=Primary] {
    backgroundColor: red
    paddingLeft: $paddingLeft

    [@hover] {
      paddingLeft: 10
      paddingRight: $paddingRight
    }
    [Mode=Dark] {
      backgroundColor: blue
      paddingLeft: $padLeft
    }
  }
  [Mode=Light] {
    backgroundColor: green
  }
}

text Label {
  lineHeight: $lineHeight

  [@hover] {
    color: red
    fontSize: $fontSize
  }

  [@minWidth=320] {
    color: green
    fontSize: $mediaFontSize
    [@hover] {
      fontSize: $mediaHoverFontSize
      color: blue
    }
  }
}`

describe('Compiling to CSS and JavaScript', () => {
  it('should return a map of files', () => {
    const { ast } = parse(file)

    expect(
      createGenerator({
        adapter: felaAdapter,
      })(ast, 'index.elo')
    ).toMatchSnapshot()
  })

  it('should return a map of files in devMode', () => {
    const { ast } = parse(file)

    expect(
      createGenerator({
        adapter: felaAdapter,
        devMode: true,
      })(ast, 'index.elo')
    ).toMatchSnapshot()
  })

  it('should use the given adapter', () => {
    const { ast } = parse(file)

    expect(
      createGenerator({ adapter: reactFelaAdapter })(ast, 'index.elo')
    ).toMatchSnapshot()
  })
})
