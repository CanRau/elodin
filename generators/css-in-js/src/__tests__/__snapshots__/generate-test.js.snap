// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiling to CSS and JavaScript should return a map of files 1`] = `
Object {
  "Button.elo.css": "._1p38iz4 {
  background-color: red;
  padding-left: 10px;
  border: 0px
}

._1p38iz4__Type-Primary {
  color: red
}

._1p38iz4__Type-Primary__Mode-Dark {
  color: blue
}

._1p38iz4__Mode-Light {
  color: green
}",
  "Button.elo.js": "import './Button.elo.css'
import { getClassNameFromVariantMap } from '@elodin/runtime'

const variantMap = {
  \\"\\": {},
  \\"__Type-Primary\\": {
    \\"Type\\": \\"Primary\\"
  },
  \\"__Type-Primary__Mode-Dark\\": {
    \\"Type\\": \\"Primary\\",
    \\"Mode\\": \\"Dark\\"
  },
  \\"__Mode-Light\\": {
    \\"Mode\\": \\"Light\\"
  }
}

export function Button(props = {}) {
  return {
    _className: getClassNameFromVariantMap('_elo_view _1p38iz4', variantMap, props),
    \\"borderWidth\\":props.borderWidth
  }
}",
  "Label.elo.css": "._1edorjp:hover {
  color: red
}",
  "Label.elo.js": "import './Label.elo.css'

export function Label(props = {}) {
  return {
    _className: '_elo_text _1edorjp',
    \\"lineHeight\\":props.lineHeight,
    \\":hover\\":{\\"fontSize\\":props.fontSize}
  }
}",
  "undefined.js": "import { Button } from './Button.elo.js'
import { Label } from './Label.elo.js'

export {
  Button,
  Label
}",
}
`;

exports[`Compiling to CSS and JavaScript should use the given adapter 1`] = `
Object {
  "Button.elo.css": "._1p38iz4 {
  background-color: red;
  padding-left: 10px;
  border: 0px
}

._1p38iz4__Type-Primary {
  color: red
}

._1p38iz4__Type-Primary__Mode-Dark {
  color: blue
}

._1p38iz4__Mode-Light {
  color: green
}",
  "Button.elo.js": "import './Button.elo.css'
import { createComponent } from 'react-fela'
import { getClassNameFromVariantMap } from '@elodin/runtime'

const variantMap = {
  \\"\\": {},
  \\"__Type-Primary\\": {
    \\"Type\\": \\"Primary\\"
  },
  \\"__Type-Primary__Mode-Dark\\": {
    \\"Type\\": \\"Primary\\",
    \\"Mode\\": \\"Dark\\"
  },
  \\"__Mode-Light\\": {
    \\"Mode\\": \\"Light\\"
  }
}

function Button(props  = {}) {
  return {
    _className: getClassNameFromVariantMap('_elo_view _1p38iz4', variantMap, props),
    \\"borderWidth\\":props.borderWidth
  }
}

export default createComponent(Button)",
  "Label.elo.css": "._1edorjp:hover {
  color: red
}",
  "Label.elo.js": "import './Label.elo.css'
import { createComponent } from 'react-fela'

function Label(props  = {}) {
  return {
    _className: '_elo_text _1edorjp',
    \\"lineHeight\\":props.lineHeight,
    \\":hover\\":{\\"fontSize\\":props.fontSize}
  }
}

export default createComponent(Label)",
  "root.js": "import { Button } from './Button.elo.js'
import { Label } from './Label.elo.js'

export {
  Button,
  Label
}",
}
`;
